File: MultidimensionalBoxIterator.py
import numpy as np

def rho(a, b):
    return max([abs(x[0]-x[1]) for x in zip(a, b)])

def minimum_dist(items, coords):
    min_d = 1e9
    for t in items:
        min_d = min(min_d, rho(coords, t))
    return min_d

class MultidimensionalBoxIterator:
    def __init__(self, dims, r):
        self.dims, self.r = dims, r
        self.size = len(dims)
        self.p_dims = [1] * (self.size + 1)

        for i in range(self.size-1, -1, -1):
            self.p_dims[i] = self.p_dims[i+1] * self.dims[i]

        self.verbose = False

    def get_number(self, coords):
        pos = 0
        for i in range(0, len(coords)):
            pos += self.p_dims[i + 1] * coords[i]
        return pos

    def get_coords(self, l):
        coords = []
        for i in range(self.size):
            coords.append((l // self.p_dims[i + 1]) % self.dims[i])
        return tuple(coords)

    def place_next(self, items, l):
        box_size = self.p_dims[0]

        while(l < box_size):
            c = self.get_coords(l)
            if(minimum_dist(items, c) >= self.r): 
                items.append(c)
                return l
            l += 1
        return -1

    def next_combination(self, items):
        M = len(items)

        while(len(items) > 0):
            l = self.get_number(items.pop()) + 1
            
            repeat = M - len(items)
            for _ in range(repeat):
                if (l := self.place_next(items, l)) < 0:
                    break
            
            if len(items) == M:
                break
            
        return items

    def run(self, M):
        l, cnt, items = 0, 0, []
        
        for i in range(M):
            if((l := self.place_next(items, l)) < 0):
                return cnt

        start = self.get_number(items[0])

        while len(items) > 0:
            cnt += 1

            if self.verbose:
                pos = self.get_number(items[0])
                if(pos != start):
                    start = pos
                    print(f'start position moved to {items[0]}')

                if self.size == 2:
                    field = np.zeros((self.dims[0], self.dims[1]))
                    print('Next combination: ')
                    for t in items:
                        field[t[0], t[1]] = 1
                    print(field)

            items = self.next_combination(items)

        return cnt

from sympy import binomial

def test_main():
    A, B, C = 8, 8, 8
    r, k = 7, 4

    m = MultidimensionalBoxIterator([A, B, C], r)
    m.verbose = True
    
    print(m.run(k), binomial(A*B*C, k))
    #print(binomial(A*B*C, k))

def test_2d(): # OEIS A061996
    ans = []
    for N in range(1, 10):
        m = MultidimensionalBoxIterator([N, N], 2)
        ans.append(m.run(3))
    print(ans)

def test_3d(): # OEIS A166540
    ans = []
    for N in range(1, 8):
        m = MultidimensionalBoxIterator([N, N, N], 2)
        ans.append(m.run(2))
    print(ans)

if __name__ == '__main__':
    #test_3d()
    test_main()

File: new/MultidimensionalBoxIterator.cpp
#include <cstdio>
#include <vector>
#include <set>
#include <chrono>
#include <string>
#include <climits>
#include <map>
#include <algorithm>
#include <iostream>

using namespace std;

typedef unsigned long long ull;

inline int rho(vector <int> &a, vector <int> &b){
    int m = -1;
    for(int i = 0; i < a.size(); ++i)
       m = max(m, abs(a[i] - b[i]));
    return m;
}

inline int minimum_dist(vector <vector <int>> &items, vector <int> &coords){
    int min_d = 1e9;
    for(auto t : items)
        min_d = min(min_d, rho(coords, t));
    return min_d;
}

class MultidimensionalBoxIterator{
    public:

    vector <int> dims;
    vector <int> p_dims;
    int r, size;
    vector <vector <int>> m_get_coords;

    MultidimensionalBoxIterator(vector <int> &dims, int r){
        this->dims = dims, this->r = r;
        this->size = dims.size();
        this->p_dims = vector <int> (this->size + 1, 1);

        for(int i = this->size - 1; i >= 0; --i)
            this->p_dims[i] = this->p_dims[i+1] * this->dims[i];

        int l = 0;
        while(l < this->p_dims[0])
            this->m_get_coords.push_back(get_coords(l++));
    }

    int get_number(vector <int> &coords){
        int pos = 0;
        for(int i = 0; i < coords.size(); ++i)
            pos += this->p_dims[i + 1] * coords[i];
        return pos;
    }

    vector <int> get_coords(int l){
        vector <int> coords;
        for(int i = 0; i < this->size; ++i)
            coords.push_back((l / this->p_dims[i + 1]) % this->dims[i]);
        return coords;
    }

    int place_next(vector <vector <int>> &items, int &l){
        int box_size = this->p_dims[0];

        while(l < box_size){
            vector <int> c = this->m_get_coords[l]; //this->get_coords(l);
            if(minimum_dist(items, c) >= this->r){
                items.push_back(c);
                return l;
            }
            ++l;
        }
        return -1;
    }

    void next_combination(vector <vector <int>> &items){
        int M = items.size();

        while(items.size() > 0){
            int l = this->get_number(items.back()) + 1;
            items.pop_back();

            int repeat = M - items.size();
            for (int k = 0; k < repeat; ++k)
                if((l = this->place_next(items, l)) < 0)
                    break;

            if(items.size() == M)
                break;
        }
    }

    ull run(int M){
        int l = 0;
        ull cnt = 0;
        vector <vector <int>> items;
        
        for(int k = 0; k < M; ++k)
            if((l = this->place_next(items, l)) < 0)
                return cnt;

        while(items.size() > 0)
            this->next_combination(items), ++cnt;
        return cnt;
    }
};

// this code is faster
ull iterate_combs(int n, int k){
    vector <int> c(k);
    ull cnt = 0;
    int i;

    for (i = 0; i < k; i++) c[i] = k - i;

    while(1){
        ++cnt;
        
        i = 0;
        if (c[i]++ < n) continue;

        for(; c[i] >= n - i;) if(++i >= k) return cnt;
        for (c[i]++; i != 0; i--) c[i-1] = c[i] + 1;
    }
    return cnt;
}

/*ull iterate_combs(int n, int k){
    vector <int> p(n);
    ull cnt = 0;

    for(int i = 0; i < n; ++i){ p[i] = (i < k); }

    do{
        ++cnt;
    }while(prev_permutation(p.begin(), p.end()));

    return cnt;
}*/

int main(void){
    vector <int> dims = {10, 8, 8};
    int r = 7, k = 5;
    auto m = MultidimensionalBoxIterator(dims, r);

    std::chrono::high_resolution_clock::time_point start, stop;

    start = chrono::high_resolution_clock::now();
    ull ans = m.run(k); 
    stop = chrono::high_resolution_clock::now();
    
    printf("Ready in %lf s.\n", chrono::duration<double, milli>(stop-start).count()/1000);

    /*start = chrono::high_resolution_clock::now();
    ull cnt = iterate_combs(m.p_dims[0], k); 
    stop = chrono::high_resolution_clock::now();
    
    printf("Ready in %lf s.\n", chrono::duration<double, milli>(stop-start).count()/1000);
    cout << ans << " " << cnt << endl;*/
    cout << ans << endl;
    return 0;
}
